File: src/main/java/xyz/chunkstories/api/entity/ai/AI.java
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.entity.ai;

import xyz.chunkstories.api.entity.Entity;

import javax.annotation.Nullable;

public abstract class AI<T extends Entity> {
	protected T entity;
	@Nullable
	protected AiTask currentTask;

	public abstract class AiTask {

		public abstract void execute();
	}

	public AI(T entity) {
		this.entity = entity;
	}

	public void tick() {
		if (currentTask != null)
			currentTask.execute();
	}

	public void setAiTask(@Nullable AiTask nextTask) {
		this.currentTask = nextTask;
	}

	@Nullable
	public AiTask currentTask() {
		return currentTask;
	}
}

File: src/main/java/xyz/chunkstories/api/entity/ai/AI.kt
//
// This file is a part of the Chunk Stories API codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.api.entity.ai

import xyz.chunkstories.api.entity.Entity

abstract class AI<E : Entity>(val entity: E) {
    var currentTask: AiTask<E> = AiTaskIdle(this)

    open fun tick() {
        currentTask.execute()
    }
}

/** Using inner classes for those was a mistake */
abstract class AiTask<E : Entity>(val ai: AI<E>) {
    val entity: E
        get() = ai.entity

    abstract fun execute()
}

class AiTaskIdle<E: Entity>(ai: AI<E>) : AiTask<E>(ai) {
    override fun execute() {
        // Do nothing (idle)
    }
}